doc/specification.md

```diff
@@ -35,62 +35,126 @@ TLS
+`ModifyRequest.election_id` MUST be non zero. When a device receives an value of 0, it should close the `Modify` RPC and set `Status.code` to `INVALID_ARGUMENT`. The election ID can only be increased monotonically by a client during a RPC session. This simplifies server implementation.
 
-### 4.1.2 Forward References
+### 4.1.3 AFTOperation
+
+RIB modification desire sent by clients are carried by a set of `AFTOperation` messages. Three types of operations are supported:
+* `ADD` - creates an entry. If the entry already exists in the specified RIB table, the `ADD` SHOULD be treated as replacing the existing entry with the entry specified in the operation.
+* `REPLACE` - replaces an existing entry in the specified RIB table. It MUST fail if the entry does not exist. A replace operation should contain all of the relevant fields, such that existing entry is completely replaced with the specified entry.
+* `DELETE` - removes an entry from the specified RIB table, it MUST fail if the entry does not exist.
+
+`AFTOperation` is identified by its `id`.  The `AFTOperation.id` should be unique per `Modify` RPC session. It's the client's responsibility to gurrantee the uniqueness during a `Modify` RPC session.
+
+#### 4.1.3.1 AFTOperation Validation
+* minimum validation.
+* `election_id` check.
+
+#### 4.1.3.2 Forward References
 
 * Ability to NACK forward references
 * Server ability for resolving forward references is not required.
 * Client's responsibility to send AFTOperations in correct order.
 
-### 4.1.3 Session Negotiation
+#### 4.1.3.3 AFTOperation Response
+
+Each AFTOperation should be responded individually.
+
+* FIB ACK vs. RIB ACK
+* When an ACK is sent to the client.
+* NACK cases:
+  * semantically invalid
+  * hardware failure
+  * missing entry for `DELETE`
+* coaelscion - must ACK every operation ID
+* acknowledging entries in the presence of other protocol routes.
+
+#### 4.1.3.4 Life cycle of an `AFTOperation`
+
+The life of an AFTOperation starts when a client creates it, and ends in the following scenarios:
+* The device failed to program the operation into RIB. // Return `FAILED`.
+* `ack_type` = `RIB_ACK`, the device programmed the operation into RIB successfully. // Return `RIB_PROGRAMMED`.
+* `ack_type` = `RIB_AND_FIB_ACK`, the device programmed the operation into FIB successfully. // Return `FIB_PROGRAMMED`.
+* `ack_type` = `RIB_AND_FIB_ACK`, the device has successfully programmed the operation into RIB but failed to program it into FIB. // Return `FIB_FAILED`. Note that this is regardless if a device is going to retry the FIB programming or not. The client can promptly send another `AFTOperations` for explicit behaviors (e.g. `ADD` for retry, and `DELETE` for stopping retry).
+* The existing gRPC session is disconnected/canceled. All pending `AFTOperations` from the client should be cancelled.
+* The device has discovered a failover of the master client (see xxx for more details).
```

Comment:
What does a failover of the master client? Does this mean a change in the elected leader?

Reply:
Yes, "a change in the elected leader" is a better wording. Done.